cpukit/score/src/schedulerstrongapa.c

```diff
@@ -3,339 +3,820 @@
+
+    /* Checks if the thread_CPU is in the affinity set of the node */
+    if ( _Processor_mask_Is_set( &strong_node->Affinity, cpu_index ) ) {
+      Per_CPU_Control *cpu = _Per_CPU_Get_by_index( cpu_index );
+
+      if ( _Per_CPU_Is_processor_online( cpu ) ) {
+        rear = rear + 1;
+        CPU[ rear ].cpu = cpu;
+        CPU[ cpu_index ].visited = true;
+        CPU[ cpu_index ].preempting_node = node;
+      }
+    }
+  }
+
+  /*
+   * This assert makes sure that there always exist an element in the
+   * Queue when we start the queue traversal.
+   */
+  _Assert( !_Processor_mask_Is_zero( &strong_node->Affinity ) );
+
+  lowest_reachable = _Scheduler_strong_APA_Get_lowest_reachable(
+                       self,
+                       front,
+                       rear,
+                       &cpu_to_preempt
+                     );
+
+  return _Scheduler_strong_APA_Do_enqueue(
+           context,
+           lowest_reachable,
+           node,
+           insert_priority,
+           cpu_to_preempt
+         );
+}
+
+static inline bool _Scheduler_strong_APA_Enqueue_scheduled(
+  Scheduler_Context *context,
+  Scheduler_Node    *node,
+  Priority_Control   insert_priority
+)
+{
+  return _Scheduler_SMP_Enqueue_scheduled(
     context,
-    the_thread,
     node,
-    _Scheduler_SMP_Extract_from_scheduled,
+    insert_priority,
+    _Scheduler_SMP_Priority_less_equal,
     _Scheduler_strong_APA_Extract_from_ready,
     _Scheduler_strong_APA_Get_highest_ready,
+    _Scheduler_strong_APA_Insert_ready,
+    _Scheduler_SMP_Insert_scheduled,
     _Scheduler_strong_APA_Move_from_ready_to_scheduled,
-    _Scheduler_SMP_Allocate_processor_exact
+    _Scheduler_strong_APA_Allocate_processor
   );
 }
 
-static bool _Scheduler_strong_APA_Enqueue(
+static inline bool _Scheduler_strong_APA_Do_ask_for_help(
   Scheduler_Context *context,
-  Scheduler_Node    *node,
-  Priority_Control   insert_priority
+  Thread_Control    *the_thread,
+  Scheduler_Node    *node
 )
 {
-  return _Scheduler_SMP_Enqueue(
+  return _Scheduler_SMP_Ask_for_help(
     context,
+    the_thread,
     node,
-    insert_priority,
     _Scheduler_SMP_Priority_less_equal,
     _Scheduler_strong_APA_Insert_ready,
     _Scheduler_SMP_Insert_scheduled,
     _Scheduler_strong_APA_Move_from_scheduled_to_ready,
-    _Scheduler_SMP_Get_lowest_scheduled,
-    _Scheduler_SMP_Allocate_processor_exact
+    _Scheduler_strong_APA_Get_lowest_scheduled,
+    _Scheduler_strong_APA_Allocate_processor
   );
 }
 
-static bool _Scheduler_strong_APA_Enqueue_scheduled(
+static inline  void  _Scheduler_strong_APA_Do_set_affinity(
   Scheduler_Context *context,
-  Scheduler_Node    *node,
-  Priority_Control  insert_priority
+  Scheduler_Node    *node_base,
+  void              *arg
 )
 {
-  return _Scheduler_SMP_Enqueue_scheduled(
+  Scheduler_strong_APA_Node *node;
+
+  node = _Scheduler_strong_APA_Node_downcast( node_base );
+  node->Affinity = *( (const Processor_mask *) arg );
```

Comment:
here should you confirm it is a non-zero affinity?

Reply:
The non-zero affinity is checked once when the node asks for help. Should I have an assert here as well?

Reply:
as long as you do it somewhere and it is well-defined behavior, then it is ok.